Zend/zend_language_parser.y

```diff
@@ -1155,7 +1155,7 @@ callable_variable:
 	|	constant '[' optional_expr ']'
 			{ $$ = zend_ast_create(ZEND_AST_DIM, $1, $3); }
 	|	dereferencable '{' expr '}'
-			{ $$ = zend_ast_create(ZEND_AST_DIM, $1, $3); }
+			{ $$ = zend_ast_create_ex(ZEND_AST_DIM, ZEND_ALTERNATIVE_ARRAY_SYNTAX, $1, $3); }
 	|	dereferencable T_OBJECT_OPERATOR property_name argument_list
```

Comment:
Wouldn't it be better to just call ``zend_error(E_DEPRECATED, "Array and string offset access syntax with curly braces is deprecated");`` directly here?
Also to cover the constant expression you want to edit line 1275
``|	T_CURLY_OPEN variable '}' { $$ = $2; }`` in ``encaps_var:``

If doing the deprecation directly it would become:
``|	T_CURLY_OPEN variable '}' { $$ = $2;  zend_error(E_DEPRECATED, "Array and string offset access syntax with curly braces is deprecated"); }``

I'm maybe missing something else but from my rapid testing this should do it

Reply:
I can be mistaken, since much time has passed, but in first approach I did it similar way. And Nikita(?) advised to move it away from parser. Unfortunately I do not remember details why.

Reply:
@rjhdby may well be possible because my change seems to hit also variables in double quotes strings ``{$var}`` in strings. So it is maybe smarter to not do it in the parser.
Will see if another change with some more testing (i.e. actually running the test suite) doesn't result in these false positives.

Reply:
The motivation for not doing it in the parser is to avoid a deprecation warning in php-ast or token_get_all in TOKEN_PARSE mode.